compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs

```diff
@@ -0,0 +1,158 @@
+        T.concat ["choice ", dot m, ":", dot t, ".", c]
+  where
+    dot = T.intercalate "."
+
+-- | Asks whether a name collision is permitted. According to the
+-- LF Spec, a name collision is only permitted when it occurs
+-- between a record type and a variant constructor defined in
+-- the same module.
+nameCollisionPermitted :: Name -> Name -> Bool
+nameCollisionPermitted a b =
+    case (a,b) of
+        (NRecordType m1 _, NVariantCon m2 _ _) -> m1 == m2
+        (NVariantCon m1 _ _, NRecordType m2 _) -> m1 == m2
+        _ -> False
+
+-- | Asks whether a name collision is forbidden.
+nameCollisionForbidden :: Name -> Name -> Bool
+nameCollisionForbidden a b = not (nameCollisionPermitted a b)
+
+-- | Fully resolved name within a package. We don't use
+-- Qualified from DA.Daml.LF.Ast because that hides collisions
+-- between module names and type names. This should only be
+-- constructed lower case in order to have case-insensitivity.
+--
+-- This corresponds to the following section of the LF spec:
+-- https://github.com/digital-asset/daml/blob/master/daml-lf/spec/daml-lf-1.rst#fully-resolved-name
+newtype FRName = FRName [T.Text]
+    deriving (Eq, Ord)
+
+-- | Turn a name into a fully resolved name.
+fullyResolve :: Name -> FRName
+fullyResolve = FRName . map T.toLower . \case
+    NModule (ModuleName m) ->
+        m
+    NRecordType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NVariantType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NEnumType (ModuleName m) (TypeConName t) ->
+        m ++ t
+    NVariantCon (ModuleName m) (TypeConName t) (VariantConName v) ->
+        m ++ t ++ [v]
+    NEnumCon (ModuleName m) (TypeConName t) (VariantConName v) ->
+        m ++ t ++ [v]
+    NField (ModuleName m) (TypeConName t) (FieldName f) ->
+        m ++ t ++ [f]
+    NChoice (ModuleName m) (TypeConName t) (ChoiceName c) ->
+        m ++ t ++ [c]
+
+-- | State of the name collision checker. This is a
+-- map from fully resolved names within a package to their
+-- original names. We update this map as we go along.
+newtype NCState = NCState (M.Map FRName [Name])
+
+-- | Initial name collision checker state.
+initialState :: NCState
+initialState = NCState M.empty
+
+-- | Try to add a name to the NCState. Returns Error only
+-- if the name results in a forbidden name collision.
+addName :: Name -> NCState -> Either Error NCState
+addName name (NCState nameMap) = do
+    let frName = fullyResolve name
+        oldNames = fromMaybe [] (M.lookup frName nameMap)
+        badNames = filter (nameCollisionForbidden name) oldNames
+    if null badNames then do
+        Right . NCState $ M.insert frName (oldNames ++ [name]) nameMap
+    else do
+        Left $ EForbiddenNameCollision
+            (displayName name)
+            (map displayName badNames)
+
+checkName :: MonadGamma m => Name -> S.StateT NCState m ()
+checkName name = do
+    ncState <- S.get
+    either throwWithContext S.put (addName name ncState)
```

Comment:
You probably wan to force the map here. Otherwise you might end up with an unevaluated `M.insert …` Probably not a big issue since you’ll force it in the next `addName` but that kind of stuff can easily change during an refactoring and turn into a proper leak.

Reply:
Good call, thanks :)